### ISSUE TITLE
How to make use of dynamic batching with Triton Python backend?

### ISSUE THREAD

**Issue by nqbao11** [triton-inference-server/server#6740]
**Description**
I would not say this is a bug but I cannot make use of the `dynamic_batching` feature with the Triton Python Backend.

**Triton Information**
Triton image: nvcr.io/nvidia/tritonserver:23.10-pyt-python-py3

**To Reproduce**
I created a dummy Python backend, which accepts `input` named tensors and returns `output` named tensors.
- model.py
```python
import logging
import triton_python_backend_utils as pb_utils
import numpy as np

logging.basicConfig(format="%(asctime)s %(message)s")
logger = logging.getLogger()
logger.setLevel(logging.INFO)

class TritonPythonModel:
    def initialize(self, args):
        logger.info("Init")

    def execute(self, requests):  
        logger.info(f"Input length: {len(requests)}")  
        responses = []
        for request in requests:
            
            input = pb_utils.get_input_tensor_by_name(request, "input").as_numpy()
            
            logger.info(f"Input shape: {input.shape}")
            
            ###
            # Perform inference task against the `input` tensor
            ###
            
            inference_response = pb_utils.InferenceResponse(
                output_tensors=[pb_utils.Tensor("output", np.asarray("Dummy output", dtype=object))]
            )
            responses.append(inference_response)
        
        return responses

    def finalize(self):
        logger.info("Cleaned")
```
- config.pbtxt
```python
name: "dummy"
backend: "python"
max_batch_size: 8

instance_group [
    {
        count: 1
        kind: KIND_GPU
    }
]

input [
    {
        name: "input"
        data_type: TYPE_STRING
        dims: [ -1 ]
    }
]

output [
    {
        name: "output"
        data_type: TYPE_STRING
        dims: [ -1 ]
    }
]

dynamic_batching {
}
```

- Start server command
```cmd
tritonserver --model-repository=/models --log-verbose=1 --model-control-mode=explicit --load-model=dummy
```
- Test 
```cmd
docker run -it --rm --net host nvcr.io/nvidia/tritonserver:23.11-py3-sdk bash
perf_analyzer -m dummy --percentile=95 --concurrency-range 1:4 --shape input:7
```
**Expected behavior**
Following [this optimization-related documentation](https://github.com/triton-inference-server/server/blob/main/docs/user_guide/optimization.md), I believe that when we enable dynamic batching, triton will automatically stack up requests to a batched input. Let's say one input has a shape of (1,7), based on the above `perf_analyzer` command, after using dynamic batch, the shape should be (x,7) with x larger than 1 and in the range of 2 to 8 - my max batch size.
However, in practice, what I receive is a list of requests, each of which has a single input.
```cmd
# Examples logging
2023-12-26 08:21:39,316 Input length: 3
2023-12-26 08:21:39,316 Input shape: (1, 6)
2023-12-26 08:21:39,316 Input shape: (1, 6)
2023-12-26 08:21:39,316 Input shape: (1, 6)
2023-12-26 08:21:39,317 Input length: 1
2023-12-26 08:21:39,317 Input shape: (1, 6)
2023-12-26 08:21:39,318 Input length: 3
2023-12-26 08:21:39,318 Input shape: (1, 6)
2023-12-26 08:21:39,318 Input shape: (1, 6)
2023-12-26 08:21:39,318 Input shape: (1, 6)
``` 
A more odd thing is that when I disable `dynamic_batching`, the above pattern of logging happens, which means `dynamic_batching` does not have any effects at all!
So please help me characterize these odd results.
1. In terms of dynamic batching, does triton batch multiple requests to a list of requests and then send it to one instance of my model, or does it stack up input of multiple requests on a single request with a larger batch size input? 
2. In case my Python backend's logic cannot handle dynamically batched inputs, how is a proper one?
Btw, happy holiday you guys!

**Comment by Tabrizian** [comment]
Dynamic batching is a feature of Triton Inference Server that allows inference requests to be combined by the server, so that a batch is created dynamically. This typically results in increased throughput.

To use dynamic batching with the Python backend in Triton, you need to understand that the Python backend is unique compared to some of the other backends (like TensorFlow, PyTorch, etc.). It doesn't currently implement the actual "batching" logic for you. It supports dynamic batching in the sense that Triton can gather the requests in the server/core and then send those requests together in a single API call to the model.

In your model's execute method, you should handle these batches of requests appropriately. For example, if your model is a using PyTorch or TensorFlow, you might need to stack the inputs along a new dimension to create a batched tensor, then pass this batched tensor through your model.

**Comment by abracadabra233** [comment]
> Dynamic batching is a feature of Triton Inference Server that allows inference requests to be combined by the server, so that a batch is created dynamically. This typically results in increased throughput.
> 
> To use dynamic batching with the Python backend in Triton, you need to understand that the Python backend is unique compared to some of the other backends (like TensorFlow, PyTorch, etc.). It doesn't currently implement the actual "batching" logic for you. It supports dynamic batching in the sense that Triton can gather the requests in the server/core and then send those requests together in a single API call to the model.
> 
> In your model's execute method, you should handle these batches of requests appropriately. For example, if your model is a using PyTorch or TensorFlow, you might need to stack the inputs along a new dimension to create a batched tensor, then pass this batched tensor through your model.

```
name: "dummy"
backend: "onnxruntime"
max_batch_size: 8

instance_group [
    {
        count: 1
        kind: KIND_GPU
    }
]

input [
    {
        name: "input"
        data_type: TYPE_FP32
        dims: [ -1 ]
    }
]

output [
    {
        name: "output"
        data_type: TYPE_FP32
        dims: [ -1 ]
    }
]

dynamic_batching {
2,4,8
}
```
When the Tensor shape in multiple requests is inconsistent, will Triton automatically complete the padding operation group the inputs in these multiple requests into one batch? This situation is very common, for example, when the ONNX backend model above receives a request with an input shape of 1 * 15 and an input shape of 3 * 20, will Triton automatically add these two inputs to 4 * 20? If triton will automatically complete, is there an interface to control the value of padding? If it cannot be completed automatically, will the dynamic_batching of triton only combine inputs with the same dimensions? So I need to add a Python backend to pad the inputs from multiple Requests into same dimensions in order to use dynamic_batching?I mainly want to know if Triton provides an interface to automatically complete the padding operation for multiple requests. Thanks

**Comment by Alonelymess** [comment]
> Dynamic batching is a feature of Triton Inference Server that allows inference requests to be combined by the server, so that a batch is created dynamically. This typically results in increased throughput.
> 
> To use dynamic batching with the Python backend in Triton, you need to understand that the Python backend is unique compared to some of the other backends (like TensorFlow, PyTorch, etc.). It doesn't currently implement the actual "batching" logic for you. It supports dynamic batching in the sense that Triton can gather the requests in the server/core and then send those requests together in a single API call to the model.
> 
> In your model's execute method, you should handle these batches of requests appropriately. For example, if your model is a using PyTorch or TensorFlow, you might need to stack the inputs along a new dimension to create a batched tensor, then pass this batched tensor through your model.

If I stack the input to process in batch, the response will be wrong. Example: 2 request concurrrently send to server, if I batch those, it will become 1 response and the server will prompt "message: Number of InferenceResponse objects do not match the number of InferenceRequest objects. InferenceRequest(s) size is:2, and InferenceResponse(s) size is:1". Pretty dumb

**Comment by nqbao11** [comment]
@Alonelymess Yes, we need to convert the output to list of "single" responses again. In their [Python Backend examples](https://github.com/triton-inference-server/python_backend/tree/main/examples), the **excute** function always have a loop over the **requests** object then append the result to **responses** object. This **responses** object will be sent by the backend all-in-one.

**Comment by Alonelymess** [comment]
> @Alonelymess Yes, we need to convert the output to list of "single" responses again. In their [Python Backend examples](https://github.com/triton-inference-server/python_backend/tree/main/examples), the **excute** function always have a loop over the **requests** object then append the result to **responses** object. This **responses** object will be sent by the backend all-in-one.

Thanks for replying me, if that is the workflow of Python backend, should we batch the request on client-side already? Suppose the request message doesn't exceed the size

**Comment by nqbao11** [comment]
I don't have answer for your question because I am still blind here. That depends on the system design, I believe. One complex use case I am thinking about is that if we do client-side batching, let's say (x, ...) and use dynamic batching, how we decouple the response into the origin batch size?

**Comment by Alonelymess** [comment]
> I don't have answer for your question because I am still blind here. That depends on the system design, I believe. One complex use case I am thinking about is that if we do client-side batching, let's say (x, ...) and use dynamic batching, how we decouple the response into the origin batch size?

If you do client-side batching, there is no need to use dynamic batching in triton. Dynamic batching in triton will gather the data that can be batched (same length) so let say if you batch (x, ...) on client-side and send 2 request contain (x, ...) each, the triton will batch them again to be (2x, ...). So if you do want to use that, you have to create a new input that indicate the index of the data so that you can decouple into original batch size
